src/backend/joanie/core/flows/order.py

```diff
@@ -208,7 +305,12 @@ def _post_transition_success(self, descriptor, source, target):  # pylint: disab
         # Only enroll user if the product has no contract to sign, otherwise we should wait
         # for the contract to be signed before enrolling the user.
         if (
-            target == enums.ORDER_STATE_VALIDATED
+            target
+            in [
+                enums.ORDER_STATE_COMPLETED,
+                enums.ORDER_STATE_FAILED_PAYMENT,
+                enums.ORDER_STATE_PENDING_PAYMENT,
+            ]
             and self.instance.product.contract_definition is None
```

Comment:
This should be trigger once. As soon as the user is able to enroll to course runs.
```suggestion
            source
            in [
                enums.ORDER_STATE_ASSIGNED,
                enums.ORDER_STATE_PENDING,
                enums.ORDER_STATE_NO_PAYMENT
            ]
            and target
            in [
                enums.ORDER_STATE_COMPLETED,
                enums.ORDER_STATE_PENDING_PAYMENT
            ]
```